### PR TITLE
re: Make `test_nearlib.sh` work from any directory

### PR DESCRIPTION
**Pull request by pmnoxx**
`scripts/test_nearlib.sh` currently doesn't work on localhost.
Before, we fix `yarn test`, let's make this script work from any directory.

See #5757

### CODE REVIEW

File: scripts/test_nearlib.sh [hunk 1]
```diff
@@ -1,12 +1,16 @@
 #!/usr/bin/env bash
 set -ex
 
+SRC_DIR=$(pwd)/$(dirname "${BASH_SOURCE[0]}")
+export REPO_DIR=$(readlink -e "$SRC_DIR/..")
+pushd "${REPO_DIR}"
```
Comment (mina86) [inline scripts/test_nearlib.sh:6]: This is really confusing. First of all, no need to `export`, just `repo_dir=whatever`. Second of all, why `pushd`? Just `cd`. And lastly, the name resolution doesn’t seem correct. I think what you want is:

    repo_dir=$(realpath -e "$(dirname "${BASH_SOURCE[0]}")/..")

File: scripts/test_nearlib.sh [hunk 2]
```diff
@@ -1,12 +1,16 @@
...
+
 export RUST_BACKTRACE=full
-./scripts/start_unittest.py --local &
+"${REPO_DIR}"/scripts/start_unittest.py --local &
```
Comment (mina86) [inline scripts/test_nearlib.sh:9]: That’s not necessary (here and in other places). With `cd` you’ve changed directory to repo dir so `./scripts/whatever` is fine.

File: scripts/test_nearlib.sh [hunk 3]
```diff
@@ -16,14 +20,13 @@ git clone https://github.com/near/create-near-app.git
 # Make sure to use local nearcore for tests
 export NODE_ENV=local
 
-SRC_DIR=$(pwd)/$(dirname "${BASH_SOURCE[0]}")
-export HOME=$SRC_DIR/../testdir
+export HOME=$(readlink -e "$REPO_DIR/../testdir")
 
 # Run near-api-js tests
-cd near-api-js
+pushd "${REPO_DIR}"/near-api-js
```
Comment (mina86) [inline scripts/test_nearlib.sh:26]: Again, `cd` is fine and also this change isn’t necessary.

Comment (pmnoxx) [reply scripts/test_nearlib.sh]: Ok, I changed it back to `cd`.

This makes reading the code easier, as I don't need to follow, in which current directory we are at the moment.

File: scripts/test_nearlib.sh (hunk 2 above)
Comment (pmnoxx) [reply scripts/test_nearlib.sh]: In different places of this script, we are either in `${repo_dir}` then in some other directory. Using absolute path, make it less error prone.

File: scripts/test_nearlib.sh (hunk 1 above)
Comment (pmnoxx) [reply scripts/test_nearlib.sh]: `push` / `popd`, show which commands need to be executed in a specific directory.

This is a good example:
```
# Run near-api-js tests
pushd "${repo_dir}"/near-api-js
yarn
yarn build
"${repo_dir}"/scripts/waitonserver.sh
popd
```

Comment (mina86) [reply scripts/test_nearlib.sh]: I’d just `cd` each time. I’d say that’s more readable than having to maintain directory stack in your head when you read the script.

**Comment by pmnoxx** [comment]
It turns out I had wrong version of `yarn` installed, I added a check. Now I have to fix the following issue
```
                                                                                             
                                                                                                                                                                                       
yarn install v1.22.17                                                                                                                                                                                              
[1/4] Resolving packages...                                                                                                                                                                                        
⠁                                                                                                                                                                                                                  
[2/4] Fetching packages...
[----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------] 0/844
error husky@7.0.4: The engine "node" is incompatible with this module. Expected version ">=12". Got "10.19.0"
error Found incompatible module.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
+ pkill -15 -P 91134

```

**Comment by pmnoxx** [comment]
Added following to ignore issues with dependencies of `near-api-js`
```
yarn config set ignore-engines true
```

File: scripts/test_nearlib.sh [hunk 4]
```diff
@@ -1,30 +1,47 @@
...
 
-export RUST_BACKTRACE=full
-./scripts/start_unittest.py --local &
+repo_dir=$(realpath -e "$(dirname "${BASH_SOURCE[0]}")/..")
+
+# Builds and starts `neard`, assumes it's running in ${repo_dir}'
+# May fail silently, leading to hard to debug CI failures
+# TODO: add option to build only and run checks, this will allow us to see those errors synchronously.
+
+pkill neard || echo "Nothing to kill"
```
[the remark this answers is not in the document]
Comment (pmnoxx) [reply scripts/test_nearlib.sh:10]: @mina86 This test is starting `neard` process.
We don't even kill it properly.

[the remark this answers is not in the document]
Comment (pmnoxx) [reply scripts/test_nearlib.sh:10]: @mina86 
I run this test, and after the test was over, "neard" still kept running.

I agree, this test should be rewritten, to do this in a better way.

[the remark this answers is not in the document]
Comment (pmnoxx) [reply scripts/test_nearlib.sh]: @mina86 Actually, it's `near`, not `neard`, which kills running after the test is over.

[the remark this answers is not in the document]
Comment (pmnoxx) [reply scripts/test_nearlib.sh]: Anyway, added TODO. Killing that process belongs to another PR.

[the remark this answers is not in the document]
Comment (mina86) [reply scripts/test_nearlib.sh]: It is starting it, but `pkill neard` will kill all processes called `neard` running on the machine including unrelated processes that user started.

The killing is supposed to happen on line 14 with `trap 'pkill -15 -P $NEAR_PID' 0`. One potential issue is that that will kill the Python script but won’t necessarily cause `neard` started by that process to die.

**Comment by pmnoxx** [comment]
````
@chefsale @mhalambek @mina86 @nikurt: /scripts/
```
@chefsale @mhalambek @mina86 @nikurt Can you take a look?

**Review by nikurt — APPROVED** [review]

File: scripts/test_nearlib.sh [hunk 5]
```diff
@@ -2,31 +2,47 @@
...
 ./scripts/requirements_check.sh
+repo_dir=$(realpath -e "$(dirname "${BASH_SOURCE[0]}")/..")
 
-export RUST_BACKTRACE=full
-./scripts/start_unittest.py --local &
+# Builds and starts `neard`, assumes it's running in ${repo_dir}'
+# May fail silently, leading to hard to debug CI failures
+# TODO: add option to build only and run checks, this will allow us to see those errors synchronously.
+
+# TODO: This test leaves `near` process running even after the test is over
```
Comment (mina86) [inline scripts/test_nearlib.sh:11]: `neard` process

File: scripts/test_nearlib.sh [hunk 6]
```diff
@@ -2,31 +2,47 @@
...
+# pkill near || echo "Nothing to kill"
+(cd "${repo_dir}"; RUST_BACKTRACE=full "${repo_dir}"/scripts/start_unittest.py --local) &
```
Comment (mina86) [inline scripts/test_nearlib.sh:13]:
```
(cd "${repo_dir}" && RUST_BACKTRACE=full ./scripts/start_unittest.py --local) &
```
`cd` followed by the use of full path is rather redundant.

File: scripts/test_nearlib.sh [hunk 7]
```diff
@@ -2,31 +2,47 @@
...
+repo_dir=$(realpath -e "$(dirname "${BASH_SOURCE[0]}")/..")
 
-export RUST_BACKTRACE=full
-./scripts/start_unittest.py --local &
+# Builds and starts `neard`, assumes it's running in ${repo_dir}'
+# May fail silently, leading to hard to debug CI failures
+# TODO: add option to build only and run checks, this will allow us to see those errors synchronously.
+
+# TODO: This test leaves `near` process running even after the test is over
+# pkill near || echo "Nothing to kill"
```
Comment (mina86) [inline scripts/test_nearlib.sh:12]: This probably should be rewritten in Python and then we can handle killing child processes with no issues.

File: scripts/test_nearlib.sh [hunk 8]
```diff
@@ -2,31 +2,47 @@
...
 # Make sure to use local nearcore for tests
 export NODE_ENV=local
-
-SRC_DIR=$(pwd)/$(dirname "${BASH_SOURCE[0]}")
-export HOME=$SRC_DIR/../testdir
+export HOME="{$repo_dir}"/testdir
 
 # Run near-api-js tests
-cd near-api-js
+pushd "${repo_dir}/near-api-js"
```
Comment (mina86) [inline scripts/test_nearlib.sh:29]: I still think `pushd` is kinda pointless. `cd` is enough.

**Review by mina86 — APPROVED** [review]

**Pull request merged**